akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala

```diff
@@ -140,15 +140,15 @@ private object PoolConductor {
               slotStates(slotIx) = slotStateAfterDisconnect(slotStates(slotIx), failed)
           }
           pull(slotIn)
-          val tryPull = nextSlot == -1
+          val wasBlocked = nextSlot == -1
           nextSlot = bestSlot()
-          if (tryPull) tryPullCtx()
+          if (wasBlocked && nextSlot != -1) pull(ctxIn)
```

Comment:
The slot selector only blocks if the currently best slot back-pressures or if there is no available slot—as soon as one becomes available it should pull again because it will be able to emit the resulting request as well.